src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceResponseTracker.cs

```diff
@@ -1,107 +1,152 @@
-﻿using System.Diagnostics;
+﻿using System;
+using System.Collections.Generic;
 using System.Linq;
 using System.Threading.Tasks;
 using Orleans.GrainDirectory;
 using Orleans.SystemTargetInterfaces;
+using OutcomeState = Orleans.Runtime.GrainDirectory.GlobalSingleInstanceResponseOutcome.OutcomeState;
 
 namespace Orleans.Runtime.GrainDirectory 
 {
-    /// <summary>
-    /// A class that encapsulates response processing logic.
-    /// It is a promise that fires once it has enough responses to make a determination.
-    /// </summary>
-    internal class GlobalSingleInstanceResponseTracker : TaskCompletionSource<GlobalSingleInstanceResponseTracker.Outcome> {
-
-        public enum Outcome {
+    internal struct GlobalSingleInstanceResponseOutcome
+    {
+        public enum OutcomeState
+        {
             Succeed,
             RemoteOwner,
             RemoteOwnerLikely,
             Inconclusive
         }
 
+        public static readonly GlobalSingleInstanceResponseOutcome Succeed = new GlobalSingleInstanceResponseOutcome(OutcomeState.Succeed, default(AddressAndTag), null);
+
+        public readonly OutcomeState State;
+        public readonly AddressAndTag RemoteOwnerAddress;
+        public readonly string RemoteOwnerCluster;
+        public GlobalSingleInstanceResponseOutcome(OutcomeState state, AddressAndTag remoteOwnerAddress, string remoteOwnerCluster)
+        {
+            this.State = state;
+            this.RemoteOwnerAddress = remoteOwnerAddress;
+            this.RemoteOwnerCluster = remoteOwnerCluster;
+        }
+    }
+
+    /// <summary>
+    /// A class that encapsulates response processing logic.
+    /// It is a promise that fires once it has enough responses to make a determination.
+    /// </summary>
+    internal class GlobalSingleInstanceResponseTracker
+    {
+        private readonly TaskCompletionSource<GlobalSingleInstanceResponseOutcome> tcs = new TaskCompletionSource<GlobalSingleInstanceResponseOutcome>();
         private readonly GrainId grain;
-        private RemoteClusterActivationResponse[] responses;
+        private readonly Task<RemoteClusterActivationResponse>[] responsePromises;
         private Logger logger;
 
-        public AddressAndTag RemoteOwner;
-        public string RemoteOwnerCluster;
-
-        public GlobalSingleInstanceResponseTracker(RemoteClusterActivationResponse[] responses, GrainId grain, Logger logger)
+        private GlobalSingleInstanceResponseTracker(Task<RemoteClusterActivationResponse>[] responsePromises, GrainId grain, Logger logger)
         {
-            this.responses = responses;
+            this.responsePromises = responsePromises;
             this.grain = grain;
             this.logger = logger;
 
             CheckIfDone();
         }
 
-        // for tracing, display outcome
-        public override string ToString()
```

Comment:
Instead of removing this `ToString()` method altogether, perhaps it should be moved to `GlobalSingleInstanceResponseOutcome`. The information it provides in the trace has been helpful to me in the past.